pkg/terraform/workspace_test.go

```diff
@@ -0,0 +1,454 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package terraform
+
+import (
+	"bytes"
+	"context"
+	"testing"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/test"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	k8sExec "k8s.io/utils/exec"
+	testingexec "k8s.io/utils/exec/testing"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	"github.com/crossplane-contrib/terrajet/pkg/resource/json"
+	tferrors "github.com/crossplane-contrib/terrajet/pkg/terraform/errors"
+)
+
+var (
+	testType              = "very-cool-type"
+	applyType             = "apply"
+	lineage               = "very-cool-lineage"
+	terraformVersion      = "1.0.10"
+	version               = 1
+	serial                = 3
+	directory             = "./testdata"
+	changeSummaryAdd      = `{"@level":"info","@message":"Plan: 1 to add, 0 to change, 0 to destroy.","@module":"terraform.ui","@timestamp":"0000-00-00T00:00:00.000000+03:00","changes":{"add":1,"change":0,"remove":0,"operation":"plan"},"type":"change_summary"}`
+	changeSummaryUpdate   = `{"@level":"info","@message":"Plan: 0 to add, 1 to change, 0 to destroy.","@module":"terraform.ui","@timestamp":"0000-00-00T00:00:00.000000+03:00","changes":{"add":0,"change":1,"remove":0,"operation":"plan"},"type":"change_summary"}`
+	changeSummaryNoAction = `{"@level":"info","@message":"Plan: 0 to add, 0 to change, 0 to destroy.","@module":"terraform.ui","@timestamp":"0000-00-00T00:00:00.000000+03:00","changes":{"add":0,"change":0,"remove":0,"operation":"plan"},"type":"change_summary"}`
+
+	state = &json.StateV4{
+		Version:          uint64(version),
+		TerraformVersion: terraformVersion,
+		Serial:           uint64(serial),
+		Lineage:          lineage,
+		RootOutputs:      map[string]json.OutputStateV4{},
+		Resources:        []json.ResourceStateV4{},
+	}
+)
+
+func newFailureWorkspace() *Workspace {
+	return NewWorkspace(directory, WithExec(&testingexec.FakeExec{
+		CommandScript: []testingexec.FakeCommandAction{
+			func(_ string, _ ...string) k8sExec.Cmd {
+				b := bytes.Buffer{}
+				b.WriteString(errBoom.Error())
+
+				return &testingexec.FakeCmd{
+					CombinedOutputScript: []testingexec.FakeAction{
+						func() ([]byte, []byte, error) {
+							return b.Bytes(), nil, errBoom
+						},
+					},
+				}
+			},
+		},
+	},
+	), WithLogger(newLogger()))
```

Comment:
In general, we initialize structs with `NewNopLogger()` in their `New*` function by default but it seems we don't do that for `NewWorkspace(..)` function. Would you mind adding that and removing these `WithLogger` statements from tests?

Reply:
In `NewWorkspace` function, `NewNopLogger` was added as a default